File: react/aaudiocore/src/services/base-audio-service.ts
import type { IAudioService } from "../contracts/audio-service";
import type { IAudioStateObserver } from "../contracts/audio-state-observer";
import type { IPlaybackStrategy } from "../contracts/playback-strategy";
import type { IPlayCommand } from "../models/play-command";
import { type AudioState, createInitialAudioState } from "../models/audio-state";
import type { PlaybackTrack } from "../models/track";
import { NormalStrategy } from "../strategies/normal-strategy";

export abstract class BaseAudioService implements IAudioService {
  protected state: AudioState = createInitialAudioState();
  protected observers = new Set<IAudioStateObserver>();
  protected playbackStrategy: IPlaybackStrategy = new NormalStrategy();
  protected queue: PlaybackTrack[] = [];
  protected currentIndex = -1;

  getState(): AudioState {
    return { ...this.state };
  }

  getQueue(): PlaybackTrack[] {
    return [...this.queue];
  }

  getCurrentIndex(): number {
    return this.currentIndex;
  }

  abstract loadMedia(trackId: string, sourceUri: string): void;
  abstract executeCommand(command: IPlayCommand): void;
  abstract seekToPosition(milliseconds: number): void;
  abstract dispose(): void;

  setQueue(queue: PlaybackTrack[], startIndex = 0): void {
    this.queue = [...queue];

    if (this.queue.length === 0) {
      this.currentIndex = -1;
      this.updateState({
        trackId: null,
        sourceUri: null,
        currentMs: 0,
        durationMs: 0
      });
      return;
    }

    this.loadTrackAt(clampIndex(startIndex, this.queue.length));
  }

  playNext(): boolean {
    const nextIndex = this.playbackStrategy.selectNext(this.currentIndex, this.queue);
    return this.loadTrackAt(nextIndex, true);
  }

  playPrevious(): boolean {
    if (this.queue.length === 0) return false;
    const previousIndex = this.currentIndex <= 0 ? this.queue.length - 1 : this.currentIndex - 1;
    return this.loadTrackAt(previousIndex, true);
  }

  setPlaybackStrategy(strategy: IPlaybackStrategy): void {
    this.playbackStrategy = strategy;
  }

  attachObserver(observer: IAudioStateObserver): void {
    this.observers.add(observer);
    observer.onAudioStateChanged(this.getState());
  }

  detachObserver(observer: IAudioStateObserver): void {
    this.observers.delete(observer);
  }

  protected updateState(patch: Partial<AudioState>): void {
    this.state = { ...this.state, ...patch };
    this.notifyObservers();
  }

  protected syncCurrentIndex(trackId: string): void {
    const index = this.queue.findIndex((track) => track.id === trackId);
    this.currentIndex = index;
  }

  protected loadTrackAt(index: number, autoplay = false): boolean {
    if (index < 0 || index >= this.queue.length) return false;

    const track = this.queue[index];
    this.currentIndex = index;
    this.loadMedia(track.id, track.sourceUri);

    if (autoplay) {
      this.playLoadedMedia();
    }

    return true;
  }

  protected abstract playLoadedMedia(): void;

  protected notifyObservers(): void {
    const snapshot = this.getState();
    this.observers.forEach((observer) => observer.onAudioStateChanged(snapshot));
  }
}

const clampIndex = (index: number, length: number): number =>
  Math.min(length - 1, Math.max(0, index));

File: winforms/aaudiocore/Services/BaseAudioService.cs
using AAudioCore.Contracts;
using AAudioCore.Models;
using AAudioCore.Strategies;

namespace AAudioCore.Services;

public abstract class BaseAudioService : IAudioService
{
    private readonly HashSet<IAudioStateObserver> observers = new();
    private readonly List<PlaybackTrack> queue = new();
    private IPlaybackStrategy playbackStrategy = new NormalStrategy();
    private AudioState state = new();
    private int currentIndex = -1;

    public AudioState GetState() => state;

    public IReadOnlyList<PlaybackTrack> GetQueue() => queue.ToList();

    public int GetCurrentIndex() => currentIndex;

    public abstract void LoadMedia(string trackId, string sourceUri);

    public abstract void ExecuteCommand(PlayCommand command);

    public abstract void SeekToPosition(double milliseconds);

    public abstract void Dispose();

    public void SetQueue(IReadOnlyList<PlaybackTrack> nextQueue, int startIndex = 0)
    {
        queue.Clear();
        queue.AddRange(nextQueue);

        if (queue.Count == 0)
        {
            currentIndex = -1;
            UpdateState(state with
            {
                TrackId = null,
                SourceUri = null,
                CurrentMs = 0,
                DurationMs = 0
            });
            return;
        }

        LoadTrackAt(Math.Min(queue.Count - 1, Math.Max(0, startIndex)));
    }

    public bool PlayNext()
    {
        var nextIndex = playbackStrategy.SelectNext(currentIndex, queue);
        return LoadTrackAt(nextIndex, true);
    }

    public bool PlayPrevious()
    {
        if (queue.Count == 0) return false;
        var previousIndex = currentIndex <= 0 ? queue.Count - 1 : currentIndex - 1;
        return LoadTrackAt(previousIndex, true);
    }

    public void SetPlaybackStrategy(IPlaybackStrategy strategy)
    {
        playbackStrategy = strategy;
    }

    public void AttachObserver(IAudioStateObserver observer)
    {
        observers.Add(observer);
        observer.OnAudioStateChanged(GetState());
    }

    public void DetachObserver(IAudioStateObserver observer)
    {
        observers.Remove(observer);
    }

    protected void UpdateState(AudioState nextState)
    {
        state = nextState;
        NotifyObservers();
    }

    protected void UpdateState(Func<AudioState, AudioState> updater)
    {
        UpdateState(updater(state));
    }

    protected void SyncCurrentIndex(string trackId)
    {
        currentIndex = queue.FindIndex(track => track.Id == trackId);
    }

    protected bool LoadTrackAt(int index, bool autoplay = false)
    {
        if (index < 0 || index >= queue.Count) return false;

        var track = queue[index];
        currentIndex = index;
        LoadMedia(track.Id, track.SourceUri);

        if (autoplay)
        {
            PlayLoadedMedia();
        }

        return true;
    }

    protected abstract void PlayLoadedMedia();

    private void NotifyObservers()
    {
        foreach (var observer in observers.ToList())
        {
            observer.OnAudioStateChanged(GetState());
        }
    }
}
